File: packages/shared/domain/collections.ts
import { z } from 'zod';
import { grpcDescriptorModeSchema, grpcMetadataSchema, grpcTlsModeSchema } from './grpc';
import {
  httpBinaryFileSchema,
  httpBodyModeSchema,
  httpCodeLanguageSchema,
  httpFormDataFieldSchema,
  httpHeaderSchema,
  httpUrlEncodedFieldSchema,
} from './http';

// P4 D4: the TypeScript mirror of Go's model.SavedRequest — the **request half** of
// httpRequestTabStateSchema, field name for field name, and nothing else.
//
// Three things it deliberately is not:
//   - Not the tab state. The four UI-only fields (requestPane, responsePane, responseView,
//     requestPaneHeight) stay out: they are per-tab furniture, and saving them into a collection
//     would make scrolling a pane mark a request dirty.
//   - Not the wire body. httpBodyWire drops disabled rows; a saved request must keep one — that
//     is the whole point of the checkbox.
//   - Not opaque, unlike tabs.state_json. Go writes this document too (on import), so Go owns the
//     type and this mirrors it — the same division httpclient.Body/HttpBodyWire already has (P3
//     D5), applied to a stored document instead of a call argument.
//
// Unlike most mirrors in this file's neighbourhood, this one **is** Zod-parsed — at exactly one
// boundary, `openCollectionRequestTab`, where the parsed result becomes tab state and a bad shape
// would break a render. That reuses TabKindDef.parseState's mechanism rather than adding a second
// trust boundary.
//
// `method` is a plain string, not httpMethodSchema's seven-member enum: Postman's own method list
// is 15 values plus any custom string (F4), and Go stores whatever the file said. The coercion to
// a method the builder can show happens at that same one boundary, not here — rejecting the whole
// document over one field would drop a request this app can otherwise open perfectly well.
export const httpSavedRequestSchema = /*#__PURE__*/ z.object({
  method: z.string().default('GET'),
  url: z.string().default(''),
  headers: /*#__PURE__*/ z.array(httpHeaderSchema).default([]),
  bodyMode: httpBodyModeSchema.default('none'),
  body: z.string().default(''),
  code: z.string().default(''),
  codeLanguage: httpCodeLanguageSchema.default('json'),
  urlEncoded: /*#__PURE__*/ z.array(httpUrlEncodedFieldSchema).default([]),
  formData: /*#__PURE__*/ z.array(httpFormDataFieldSchema).default([]),
  binaryFile: httpBinaryFileSchema,
});
export type HttpSavedRequest = z.infer<typeof httpSavedRequestSchema>;

/** An empty saved request — what a brand-new collection row starts as. */
export function defaultHttpSavedRequest(): HttpSavedRequest {
  return httpSavedRequestSchema.parse({});
}

// P11 D12: the TypeScript mirror of Go's model.SavedGrpcRequest — the request half of
// grpcRequestTabStateSchema (domain/grpc.ts), field name for field name, following
// httpSavedRequestSchema's own precedent exactly: not the tab state (the four UI-only fields stay
// out), and Go owns the type since it writes this document too.
export const httpSavedGrpcRequestSchema = /*#__PURE__*/ z.object({
  target: z.string().default(''),
  tlsMode: grpcTlsModeSchema.default('tls'),
  caFile: z.string().default(''),
  serverName: z.string().default(''),
  descriptorMode: grpcDescriptorModeSchema.default('reflection'),
  protoPath: z.string().default(''),
  importPaths: /*#__PURE__*/ z.array(z.string()).default([]),
  service: z.string().default(''),
  method: z.string().default(''),
  message: z.string().default(''),
  metadata: /*#__PURE__*/ z.array(grpcMetadataSchema).default([]),
});
export type HttpSavedGrpcRequest = z.infer<typeof httpSavedGrpcRequestSchema>;

/** An empty saved gRPC request — what a brand-new collection row starts as. */
export function defaultHttpSavedGrpcRequest(): HttpSavedGrpcRequest {
  return httpSavedGrpcRequestSchema.parse({});
}

// The two flat row shapes CollectionsService.List answers with (D11). Mirrors of Go's
// model.Collection / model.CollectionItem, typed rather than validated — the renderer builds the
// tree from these, the same way TreeService.Children returns flat nodes.
export interface CollectionSummary {
  id: string;
  name: string;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

/** `method` and `url` are denormalized out of the saved request so the tree renders a method chip
 *  and searches URLs without reading a body; both are '' for a folder. `protocol` (P11 D12) is
 *  structural sibling data, not a third `kind` value — `kind` stays 'folder' vs. a leaf, and
 *  `protocol` says which document shape `request_json` holds for a leaf: 'http' ->
 *  model.SavedRequest, 'grpc' -> model.SavedGrpcRequest. Meaningless for a folder, defaulted to
 *  'http' there. */
export interface CollectionItemSummary {
  id: string;
  collectionId: string;
  parentId: string | null;
  kind: CollectionItemKind;
  name: string;
  sortOrder: number;
  method: string;
  url: string;
  protocol: CollectionItemProtocol;
  createdAt: string;
  updatedAt: string;
}

export const COLLECTION_ITEM_KINDS = ['folder', 'request'] as const;
export type CollectionItemKind = (typeof COLLECTION_ITEM_KINDS)[number];

export const COLLECTION_ITEM_PROTOCOLS = ['http', 'grpc'] as const;
export type CollectionItemProtocol = (typeof COLLECTION_ITEM_PROTOCOLS)[number];

/** Which of the two tables a tree row lives in — the renderer passes this rather than knowing. */
export type CollectionTarget = 'collection' | 'item';

// P4 D12: the import report is part of the feature, not decoration. Every warning kind is a case
// where the app quietly does something other than what the file says, and the alternative to
// telling the user is letting them find out from a 401 or an E_BAD_REQUEST minutes later.
export interface ImportWarning {
  kind: string;
  count: number;
  detail: string;
}

export interface ImportReport {
  collectionId: string;
  name: string;
  folders: number;
  requests: number;
  warnings: ImportWarning[];
}

File: apps/kira-studio/internal/storage/model/collections.go
package model

import "fmt"

// P4 D4: model.SavedRequest is what a saved collection request *is* — deliberately field-identical
// to the request half of packages/shared/domain/http.ts's httpRequestTabStateSchema, so the
// renderer can spread it straight into tab state. It is NOT tabs.state_json (the four UI-only
// pane fields stay out — saving them would make scrolling a pane mark a request dirty) and it is
// NOT httpclient.Body (the wire body drops disabled rows, which is the one thing a saved request
// must keep).
//
// Method is a plain string, not a closed enum: Postman's own method list is 15 values plus any
// custom string (F4), and a value this app's builder cannot show is coerced at exactly one
// renderer boundary, not silently rejected at the storage layer.
type SavedRequest struct {
	Method       string           `json:"method"`
	URL          string           `json:"url"`
	Headers      []SavedHeader    `json:"headers"`
	BodyMode     string           `json:"bodyMode"`
	Body         string           `json:"body"` // the `raw` mode's own buffer
	Code         string           `json:"code"` // the `code` mode's own buffer
	CodeLanguage string           `json:"codeLanguage"`
	URLEncoded   []SavedField     `json:"urlEncoded"`
	FormData     []SavedFormField `json:"formData"`
	BinaryFile   *SavedFile       `json:"binaryFile"`
}

// SavedHeader is one header row. `enabled` has no wire counterpart — it exists so the builder can
// keep an unchecked row instead of deleting it, which is the whole point of the checkbox.
type SavedHeader struct {
	Name    string `json:"name"`
	Value   string `json:"value"`
	Enabled bool   `json:"enabled"`
}

// SavedField is one urlencoded row.
type SavedField struct {
	Name    string `json:"name"`
	Value   string `json:"value"`
	Enabled bool   `json:"enabled"`
}

// SavedFormField is one form-data row. Kind == "file" means Path is a local path (never bytes)
// and Value is ignored; FileName/FileSize let the builder render `report.csv (1.2 MB)` with no
// round trip back to disk.
type SavedFormField struct {
	Name        string `json:"name"`
	Kind        string `json:"kind"`
	Value       string `json:"value"`
	Path        string `json:"path"`
	FileName    string `json:"fileName"`
	FileSize    int64  `json:"fileSize"`
	ContentType string `json:"contentType"`
	Enabled     bool   `json:"enabled"`
}

// SavedFile is the binary (Postman `file`) body's one chosen file — path only, never bytes.
type SavedFile struct {
	Path string `json:"path"`
	Name string `json:"name"`
	Size int64  `json:"size"`
}

// savedBodyModes mirrors HTTP_BODY_MODES (packages/shared/domain/http.ts) and
// httpclient.validBodyModes. Kept here rather than imported so this package keeps its
// standard-library-only dependency set (internal/postman imports it, and a cycle through
// httpclient would be gratuitous).
var savedBodyModes = map[string]bool{
	"none": true, "raw": true, "code": true, "urlencoded": true, "formdata": true, "file": true,
}

// savedCodeLanguages mirrors CODE_LANGUAGES (packages/shared/domain/http.ts).
var savedCodeLanguages = map[string]bool{
	"javascript": true, "json": true, "html": true, "xml": true,
}

// Validate checks what SQL cannot. A row failing it is dropped and logged on read
// (repos/saved_queries.go's posture) and refused on write (repos/tabs.go's).
func (r SavedRequest) Validate() error {
	if r.Method == "" {
		return fmt.Errorf("model: saved request: method is required")
	}
	if !savedBodyModes[r.BodyMode] {
		return fmt.Errorf("model: saved request: unrecognised body mode %q", r.BodyMode)
	}
	if !savedCodeLanguages[r.CodeLanguage] {
		return fmt.Errorf("model: saved request: unrecognised code language %q", r.CodeLanguage)
	}
	return nil
}
